segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java

```diff
@@ -840,28 +840,22 @@ private LogMetadata updateMetadata(LogMetadata currentMetadata, LedgerHandle new
     private void persistMetadata(LogMetadata metadata, boolean create) throws DurableDataLogException {
         try {
             byte[] serializedMetadata = LogMetadata.SERIALIZER.serialize(metadata).getCopy();
-            Stat result;
-            if (create) {
-                result = new Stat();
-                this.zkClient.create()
-                             .creatingParentsIfNeeded()
-                             .storingStatIn(result)
-                             .forPath(this.logNodePath, serializedMetadata);
-            } else {
-                result = this.zkClient.setData()
-                                      .withVersion(metadata.getUpdateVersion())
-                                      .forPath(this.logNodePath, serializedMetadata);
-
-            }
-
+            Stat result = create
+                    ? createZkMetadata(serializedMetadata)
+                    : updateZkMetadata(serializedMetadata, metadata.getUpdateVersion());
             metadata.withUpdateVersion(result.getVersion());
         } catch (KeeperException.NodeExistsException | KeeperException.BadVersionException keeperEx) {
-            // We were fenced out. Clean up and throw appropriate exception.
-            throw new DataLogWriterNotPrimaryException(
-                    String.format("Unable to acquire exclusive write lock for log (path = '%s%s').", this.zkClient.getNamespace(), this.logNodePath),
-                    keeperEx);
+            if (reconcileMetadata(metadata)) {
```

Comment:
If we reached this point and we get a `BadVersionException` for an update that has been actually committed (origin of the issue), this PR can fix the problem (by not throwing, so keeping the BK ledgers along with the new ledger metadata) in the case that the `zkClient` has a session established with the server. However, as the origin of this issue itself is closely related to the instability of the network, it may be the case that the `reconcileMetadata()` fails as it is not able to load the current state (whatever it is) from Zookeeper (session may be lost or suspended). If I'm not mistaken, this will leave us in the same situation: the Zookeeper session will be re-established sooner or later, then the `initialize()` method will get executed upon a container recovery, and the same symptom will appear when executing `Ledgers.fenceOut()` (the metadata for a ledger is in ZK, but the ledger is not in BK), so the container will keep restarting indefinitely. The question is: what do we do at this point in the code if we cannot know what is the state of the data in Zookeeper? Should we try to reconcile when the container restarts again?